test/integration/network/services_test.go

```diff
@@ -0,0 +1,455 @@
+		dualStack       bool
+	}{
+		{
+			name:            "Single Stack IPv4",
+			serviceCIDR:     "10.0.0.0/16",
+			defaultIPFamily: v1.IPv4Protocol,
+			dualStack:       false,
+		},
+		{
+			name:            "Single Stack IPv6",
+			serviceCIDR:     "2001:db8:1::/48",
+			defaultIPFamily: v1.IPv6Protocol,
+			dualStack:       false,
+		},
+		{
+			name:            "DualStack enabled and IPv4only",
+			serviceCIDR:     "10.0.0.0/16",
+			defaultIPFamily: v1.IPv4Protocol,
+			dualStack:       true,
+		},
+		{
+			name:            "DualStack enabled and IPv6only",
+			serviceCIDR:     "2001:db8:1::/48",
+			defaultIPFamily: v1.IPv6Protocol,
+			dualStack:       true,
+		},
+		{
+			name:            "DualStack enabled and IPv4 IPv6",
+			serviceCIDR:     "10.0.0.0/16,2001:db8:1::/48",
+			defaultIPFamily: v1.IPv4Protocol,
+			dualStack:       true,
+		},
+		{
+			name:            "DualStack enabled and IPv6 IPv4",
+			serviceCIDR:     "2001:db8:1::/48,10.0.0.0/16",
+			defaultIPFamily: v1.IPv6Protocol,
+			dualStack:       true,
+		},
+		{
+			name:            "DualStack enabled and IPv6 IPv4 IPv6",
+			serviceCIDR:     "2001:db8:1::/48,10.0.0.0/16,fd00:192:168::/48",
+			defaultIPFamily: v1.IPv6Protocol,
+			dualStack:       true,
+		},
+	}
+
+	for _, tc := range testcases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			etcd := framework.SharedEtcd()
+			// cleanup the registry storage
+			defer registry.CleanupStorage()
+			// start a kube-apiserver
+			server := kubeapiservertesting.StartTestServerOrDie(t, nil, []string{
+				"--service-cluster-ip-range", tc.serviceCIDR,
+				"--advertise-address", "10.0.0.1",
+				"--feature-gates", fmt.Sprintf("IPv6DualStack=%t", tc.dualStack),
+			}, etcd)
+			// create a client
+			client, err := kubernetes.NewForConfig(server.ClientConfig)
+			if err != nil {
+				t.Errorf("error creating client: %v", err)
+			}
+
+			// verify client is working
+			if err := wait.PollImmediate(5*time.Second, 2*time.Minute, func() (bool, error) {
+				_, err = client.CoreV1().Endpoints("default").Get("kubernetes", metav1.GetOptions{})
+				if err != nil {
+					t.Logf("error fetching endpoints: %v", err)
+					return false, nil
+				}
+				return true, nil
+			}); err != nil {
+				t.Errorf("server without enabled endpoints failed to register: %v", err)
+			}
+
+			// Create a service ClusterIP without IPFamily so it's created with the default IP family
+			svc := createService("svc-transition", nil, v1.ServiceTypeClusterIP)
+			if _, err := client.CoreV1().Services(metav1.NamespaceDefault).Create(svc); err != nil {
+				t.Errorf("unexpected error text: %v", err)
+			}
+
+			// Transition the service over all the types of services and check that the IPFamily is inmutable
+			ipFamily := tc.defaultIPFamily
+
+			svcType := []v1.ServiceType{
+				v1.ServiceTypeExternalName,
+				v1.ServiceTypeNodePort,
+				v1.ServiceTypeLoadBalancer,
+				v1.ServiceTypeClusterIP,
+				v1.ServiceTypeClusterIP, // no-op
+			}
```

Comment:
we need a second dimension that simulates a client that is unaware of the new IPFamily field and drops it (sets it to nil) on every update